### ISSUE TITLE
RequestContext not populated with context processors on async request

### ISSUE THREAD

**Issue by larrybotha** [EmilStenstrom/django-components#634]
Component's rendered via an HTTP request that use `RequestContext` are not populated with values defined in `context_processors`

---

As per [Django's `RequestContext` docs](https://docs.djangoproject.com/en/5.1/ref/templates/api/#django.template.RequestContext), `RequestContext` differs from `Context` in that it makes the configured `context_processors` available when rendering templates

The context processors are added to context when `RequestContext.bind_template` is called, and things like `csrf_token` are then made available in templates

Since at least `django-components@0.95` (potentially since the introduction of the template monkey-patching) the context processors are not added to the context because [_monkeypatch_template._template_render](https://github.com/EmilStenstrom/django-components/blob/0cfc40231bed770cbe00256b2da0b9a9d2f37828/src/django_components/component.py#L838) doesn't call [`context.bind_template` for `RequestContext`](https://github.com/django/django/blob/64443f555f362ddf0c9094f8531c3da72a753827/django/template/context.py#L247) when the component is rendered via a network request. Because the component is not rendered as the child of a larger context, as when rendering a full page, there is no outer context available to the component, as when rendered during a 'normal' full-page request

## Demo

- https://replit.com/@larrybotha1/Django-Components-088#django_project/components/thing/index.html

## Steps to reproduce

- ensure `'django.contrib.auth.context_processors.auth'` is included in context processors in settings
- create a component that is rendered via a network request
- pass `context=RequestContext(request)` to `render_to_response`
- make a request for the component, rendering `{{ csrf_token }}`

## Expected

- `{{ csrf_token }}` is rendered when the component is rendered via a network request

## Actual

- `{{ csrf_token }}` is not available

## Possible solutions

I haven't yet worked out how this could be solved. The difference here is in what `Context.bind_template` and `RequestContext.bind_template` do. `Context` does little except yield the template, whereas `RequestContext` augments the context.

`django-components` does the correct thing in that it evaluates whether `template is None` before calling `bind_template`, so the problem may stem from [where `Component.template` is assigned](https://github.com/EmilStenstrom/django-components/blob/0cfc40231bed770cbe00256b2da0b9a9d2f37828/src/django_components/component.py#L524) - should `template` never be assigned directly, and only via `bind_template`?

## Workarounds

- instantiate `RequestContext` with the context processors manually

EDIT: improved clarity of description of problem

**Comment by JuroOravec** [comment]
@larrybotha Thanks for the report. Looks like this issue is probably present since https://github.com/EmilStenstrom/django-components/releases/tag/0.89 and https://github.com/EmilStenstrom/django-components/pull/582.

Not sure if I remember correctly, but I think I didn't use `bind_template` because it was raising some error about template or engine not being present or something...

I still don't fully grasp the workings of Context / RenderContext, but I wonder, in `Template.render`, Django has this:

```py
def render(self, context):
    "Display stage -- can be called many times"
    with context.render_context.push_state(self):
        if context.template is None:
            with context.bind_template(self):
                context.template_name = self.name
                return self._render(context)
        else:
            return self._render(context)
```

So my guess would be to:
- Try using `context.bind_template`, see what breaks, and make changes so we can call `bind_templates` properly
- Maybe we need to call also `with context.render_context.push_state(self):` for the context to be populated?

**Comment by JuroOravec** [comment]
Ok, @larrybotha I got the demo repo working by wrapping the rest of the `_render_impl()` body in `with context.bind_template(template):` if `context.template` is `None`.

<img width="572" alt="Screenshot 2024-09-02 at 18 23 26" src="https://github.com/user-attachments/assets/a04dea2f-9195-44a5-bb90-a8606c282180">

I'll make a test case to cover this later today or tomorrow, and then it's good to go!

**Comment by larrybotha** [comment]
@JuroOravec you're a machine

Ye, I ended up wrapping calls to `render_to_response` inside the context manager so that the template would bind before getting to `_render_impl`, so sounds about right

Thanks for taking the time from other issues and work to look into this!

**Comment by JuroOravec** [comment]
@larrybotha This has now been released in [0.96](https://github.com/EmilStenstrom/django-components/releases/tag/0.96) :)
